File: src/languages/typescript/assets/callee/dispatch.ts
/* {{IMPORTS}} */

import {
  AcknowledgeMessage,
  Bridge,
  Message,
  Parameter,
  SendMessage,
} from "./bridge.ts";

const instanceRegistry = new Map<string, unknown>();

function resolveParameterValue(param: Parameter): unknown {
  if (param.kind === "reference") {
    const instance = instanceRegistry.get(param.value);
    if (instance === undefined) {
      throw new Error(`Instance ${param.value} not found`);
    }
    return instance;
  }
  return param.value;
}

export function dispatchMessage(message: Message, bridge: Bridge) {
  message.match({
    call(message) {
      const positionalParameters = message.positionalParameters.map(resolveParameterValue);
      const namedParameters = new Map<string, unknown>();
      for (const [key, param] of message.namedParameters.entries()) {
        namedParameters.set(key, resolveParameterValue(param));
      }
      const result = dispatchFunction(
        message.modulePath,
        message.callee,
        message.returnSink,
        positionalParameters,
        namedParameters,
      );
      bridge.send(
        new SendMessage({
          reference: message.returnSink,
          value: result,
        }),
      );
    },
    method(message) {
      const instance = instanceRegistry.get(message.calledReference);
      if (instance === undefined) {
        throw new Error(`Instance ${message.calledReference} not found`);
      }
      const positionalParameters = message.positionalParameters.map(resolveParameterValue);
      const namedParameters = new Map<string, unknown>();
      for (const [key, param] of message.namedParameters.entries()) {
        namedParameters.set(key, resolveParameterValue(param));
      }
      const result = dispatchMethod(
        instance,
        message.methodName,
        positionalParameters,
        namedParameters,
        message.returnSink,
      );
      bridge.send(
        new SendMessage({
          reference: message.returnSink,
          value: result,
        }),
      );
    },
    update(message) {
      const instance = instanceRegistry.get(message.parent);
      if (instance === undefined) {
        throw new Error(`Instance ${message.parent} not found`);
      }
      if (instance === null || typeof instance !== "object") {
        throw new Error(`Invalid instance for update ${message.accessor}`);
      }
      (instance as Record<string, unknown>)[message.accessor] = resolveParameterValue(message.value);
      bridge.send(new AcknowledgeMessage({ reference: message.acknowledgeSink }));
    },
    request(message) {
      const instance = instanceRegistry.get(message.parent);
      const result = handleRequest(instance, message.accessor);
      const response = new SendMessage({
        reference: message.valueSink,
        value: result,
      });
      bridge.send(response);
    },
    send(_message) {
      console.warn("The callee cannot handle incoming send messages.");
    },
    error(_message) {
      console.warn("The callee cannot handle incoming error messages.");
    },
    drop(message) {
      instanceRegistry.delete(message.reference);
    },
  });
}

export function dispatchFunction(
  modulePath: string,
  callee: { kind: "function"; name: string } | { kind: "staticMethod"; typeName: string; methodName: string },
  returnSink: string,
  positionalParameters: unknown[],
  namedParameters: Map<string, unknown> = new Map(),
): Parameter {
  switch (callee.kind) {
    case "function": {
      switch (modulePath) {
/* {{FUNCTION_CASES}} */
      }
      break;
    }
    case "staticMethod": {
      throw new Error(`Static method not found: ${modulePath}.${callee.typeName}.${callee.methodName}`);
    }
  }

  throw new Error(`Function ${modulePath}.${"name" in callee ? callee.name : `${callee.typeName}.${callee.methodName}`} not found`);
}

function dispatchMethod(
  instance: unknown,
  methodName: string,
  positionalParameters: unknown[],
  namedParameters: Map<string, unknown> = new Map(),
  returnSink: string,
): Parameter {
/* {{METHOD_CASES}} */
  throw new Error(`Method ${methodName} not found`);
}

function handleRequest(
  parent: unknown,
  accessor: string,
): Parameter {
  if (parent === null || typeof parent !== "object") {
    throw new Error(`Invalid parent for request ${accessor}`);
  }
/* {{REQUEST_CASES}} */
  throw new Error(`Property ${accessor} not found`);
}

File: src/languages/python/assets/callee/dispatch.py
from __future__ import annotations

import importlib
import uuid as uuidlib
from typing import Any

from ffi.bridge import (
    AcknowledgeMessage,
    CallMessage,
    DropMessage,
    ErrorMessage,
    MethodMessage,
    Parameter,
    RequestMessage,
    SendMessage,
    UpdateMessage,
    parameter_to_python,
    python_to_parameter,
)
_LIBRARY = importlib.import_module("library.index")
_INSTANCE_REGISTRY: dict[str, Any] = {}

def _store_instance(value: Any) -> str:
    reference = str(uuidlib.uuid4())
    _INSTANCE_REGISTRY[reference] = value
    return reference


def _store_instance_with_reference(reference: str, value: Any) -> str:
    _INSTANCE_REGISTRY[reference] = value
    return reference

def _result_to_parameter(result: Any, reference: str | None = None) -> Parameter:
    if result is None:
        return Parameter(kind="string", value="undefined")
    if isinstance(result, Parameter):
        return result
    if reference is not None:
        return Parameter(kind="reference", value=_store_instance_with_reference(reference, result))
    if isinstance(result, bool):
        return Parameter(kind="boolean", value=result)
    if isinstance(result, int) and not isinstance(result, bool):
        return Parameter(kind="integer", value=result)
    if isinstance(result, float):
        return Parameter(kind="float", value=result)
    if isinstance(result, str):
        return Parameter(kind="string", value=result)
    if hasattr(result, "uuid"):
        instance_reference = str(getattr(result, "uuid"))
        _INSTANCE_REGISTRY[instance_reference] = result
        return Parameter(kind="reference", value=instance_reference)
    return Parameter(kind="reference", value=_store_instance(result))


def _is_constructor_call(callee: Any, target: Any) -> bool:
    return hasattr(callee, "name") and isinstance(target, type)

def _resolve_target(callee: Any) -> Any:
    if hasattr(callee, "name"):
        return getattr(_LIBRARY, callee.name)
    if hasattr(callee, "type_name") and hasattr(callee, "method_name"):
        return getattr(getattr(_LIBRARY, callee.type_name), callee.method_name)
    raise TypeError(f"Unsupported call target: {type(callee)!r}")

def _invoke_target(
    target: Any,
    positional_parameters: list[Parameter],
    named_parameters: dict[str, Parameter],
    reference: str | None = None,
) -> Parameter:
    positional = [parameter_to_python(parameter) for parameter in positional_parameters]
    named = {name: parameter_to_python(parameter) for name, parameter in named_parameters.items()}
    result = target(*positional, **named)
    return _result_to_parameter(result, reference)

def dispatch_message(message: Any) -> Any:
{{HANDLERS}}
    raise TypeError(f"Unsupported message: {type(message)!r}")
